src/app/_lib/_types/types.ts

```diff
@@ -0,0 +1,58 @@
+import { z } from "zod";
+import { registerBodySchema, userBodySchema } from "./schemas";
+
+export type FormInput = {
+  label: string;
+  type: string;
+  id: string;
+  placeholder: string;
+};
+
+export type FormData_t = {
+  title: string;
+  button: string;
+};
+
+export type RegisterInputs = {
+  email: string;
+  username: string;
+  password: string;
+};
+
+export type RegisterBody = z.infer<typeof registerBodySchema>;
+
+export type InputsEdited = {
+  email: false;
+  username: false;
+  password: false;
+};
+
+export type ServerResponse = {
+  success: boolean;
+  message: string;
+};
+
+export type User = z.infer<typeof userBodySchema> | undefined
+
+export type AuthResponse = {
+  success: boolean;
+  message: string;
+  user: User;
+};
+
+export type AuthAction = (
+  state: any,
+  formData: FormData
+) => Promise<AuthResponse>;
+
+export type ServerAction = (
+  state: any,
+  formData: FormData
+) => Promise<ServerResponse>;
+
+export type FormMode = "register" | "login" | "admin";
+
+export type AsyncFunctionResult<T extends {}> = {
+    success: boolean;
+} & T;
```

Comment:
This typescript wizardry allows me to append a `success: boolean` property on any async query. For example, on the async function `userAlreadyExists`, I could set its return type to `Promise<AsyncFunctionResult<{ userExists: boolean }>>`
In English, this means it returns a Promise that resolves to type AsyncFunctionResult, which is a combination of the { success: boolean } object and the newly specified { userExists: boolean } object. The object this function returns would be { success: boolean, userExists: boolean }